pkg/cloudprovider/openstack/openstack.go

```diff
@@ -630,31 +631,43 @@ func (lb *LoadBalancer) UpdateTCPLoadBalancer(name, region string, hosts []strin
 	return nil
 }
 
-func (lb *LoadBalancer) DeleteTCPLoadBalancer(name, region string) error {
-	glog.V(4).Infof("DeleteTCPLoadBalancer(%v, %v)", name, region)
+func (lb *LoadBalancer) EnsureTCPLoadBalancerDeleted(name, region string) error {
+	glog.V(4).Infof("EnsureTCPLoadBalancerDeleted(%v, %v)", name, region)
 
-	vip, err := getVipByName(lb.network, name)
-	if err != nil {
-		return err
+	// It's ok if the vip doesn't exist, as we may still need to delete the pool.
+	vip, vipErr := getVipByName(lb.network, name)
+	if vipErr != nil && vipErr != ErrNotFound {
+		return vipErr
 	}
+	vipExists := (vipErr == nil)
 
-	pool, err := pools.Get(lb.network, vip.PoolID).Extract()
-	if err != nil {
-		return err
+	// It's ok if the pool doesn't exist, as we may still need to delete the vip
+	// (although I don't believe the system should ever be in that state).
+	pool, poolErr := pools.Get(lb.network, vip.PoolID).Extract()
+	if poolErr != nil {
+		detailedErr, ok := poolErr.(*gophercloud.UnexpectedResponseCodeError)
+		if !ok || detailedErr.Actual != http.StatusNotFound {
+			return poolErr
+		}
 	}
+	poolExists := (poolErr == nil)
 
 	// Have to delete VIP before pool can be deleted
-	err = vips.Delete(lb.network, vip.ID).ExtractErr()
-	if err != nil {
-		return err
+	if vipExists {
+		err := vips.Delete(lb.network, vip.ID).ExtractErr()
+		if err != nil {
+			return err
+		}
 	}
 
 	// Ignore errors for everything following here
 
-	for _, monId := range pool.MonitorIDs {
-		pools.DisassociateMonitor(lb.network, pool.ID, monId)
```

Comment:
Huh, it looks like I'm not deleting the monitor object at this point - that's a bug.